vespajlib/src/main/java/com/yahoo/slime/ArrayValue.java

```diff
@@ -6,43 +6,116 @@
  */
 final class ArrayValue extends Value {
 
-    private int capacity = 16;
+    static final int initial_capacity = 16;
+
+    private interface Impl {
+        public void prepareFor(ArrayValue self, Type type);
+        public Value add(Value value, int used);
+        public Value get(int index);
+    }
+
+    private static final class EmptyImpl implements Impl {
```

Comment:
This has no state and could be a singleton. Then you avoid creating a dummy every time.

Reply:
good point